### PR TITLE
feat(testing): support for vscode-jest integration.

### LINKED ISSUE #1190: [BUG] New project instructions don't work
I followed the [official docs](https://nx.dev/tutorial/01-create-application) to create a new app.

Environment (latest of everything):
nx 7.7.2
angular-cli 7.3.1
node 10.15.3
npm 6.9.0
ubuntu 18
vscode 1.32.3

Repro:
````
sudo npm install --global @angular/cli@7.3.6     # latest
npx create-nx-workspace myworkspace              # choose empty workspace
cd myworkspace
ng g application myapp                           # choose angular, jest, cypress
npm install

npm run test                                     # succeeds
ng test                                          # succeeds
````

Problem: running tests in IDE, or using jest in terminal, fails:
- From vscode: open vscode, then open the `/apps/myapp/src/app/app.component.spec.ts`
- From terminal: `node_modules/.bin/jest apps/myapp/src/app/app.component.spec.ts --testNamePattern="AppComponent"`

In both cases error is
>`"Zone is needed for the async() test helper but could not be found. Please make sure that your environment includes zone.js/dist/zone.js"`

This is significant because the IDE tooling and various test extensions use the jest approach to find and run tests (rather than `npm run test` / `ng test`).

### PR DESCRIPTION
**Pull request by Domratchev**
Testing in VSCode is possible using vscode-jest extension with the following configuration in launch.json:
    {
      "name": "vscode-jest-tests",
      "type": "node",
      "request": "launch",
      "program": "${workspaceFolder}/node_modules/@angular/cli/bin/ng",
      "args": ["test", "--run-in-band", "--no-code-coverage", "--path-to-file-to-test='${fileDirname}'", "--file-to-test"],
      "cwd": "${workspaceFolder}",
      "console": "integratedTerminal",
      "internalConsoleOptions": "neverOpen",
      "trace": "all"
    }

Might as well run individual tests from the command line using `ng test` command with `--file-to-test` and `--path-to-file-to-test`.

The `--path-to-file-to-test` might appear redundant, but it's really helpful with vscode-jest integration, as that extension passes in only file name, without the path.

Fixes https://github.com/nrwl/nx/issues/1190

### CODE REVIEW

**Comment by Domratchev** [comment]
If you suggest different option names, e.g. --test-file & --dest-dir, I might as well do that.

**Comment by Domratchev** [comment]
With new options the launch.json config is this:
    {
      "name": "vscode-jest-tests",
      "type": "node",
      "request": "launch",
      "program": "${workspaceFolder}/node_modules/@angular/cli/bin/ng",
      "args": ["test", "--run-in-band", "--no-code-coverage", "--test-directory='${fileDirname}'", "--test-file"],
      "cwd": "${workspaceFolder}",
      "console": "integratedTerminal",
      "internalConsoleOptions": "neverOpen",
      "trace": "all"
    }

**Comment by FrozenPandaz** [comment]
Hi can we approach this in a different way?

Can we keep the options we currently have but then have a way to pass additional options to jest?

We are open to any new options you want but let's do them 1 at a time so we can consider each one.

Something like `--file` is a good start I think that one would be useful for sure.

**Comment by Domratchev** [comment]
Hi Jason, thanks for the suggestion.
Definitely can do, but conceptually, why would you want to interfere in the Jest builder with what Jest itself wants/understands? Having JSON schema around is sufficient for options validation, albeit already overly intrusive (e.g. I could have implemented defaults in it, but deliberately chosen not to, 'cause Jest would do the same).
Again, if you are certain that you want to be fully in control of all the options, I can do it quickly.
At the very minimum these options need to be supported for vscode-jest integration:
"--run-in-band", "--code-coverage" (this on is optional), "--test-name-pattern", "--test-directory" and "--test-file".
Thanks,
Alexei

**Comment by Domratchev** [comment]
@FrozenPandaz Done.

**Comment by jasedwards** [comment]
Is there a separate issue for adding other options?  My company uses subversion instead of git which means watch does not work for us.  I would still like to keep the session live which is why I would need watchAll.  Was not sure if we should open a new issue for each option needed.

File: packages/builders/src/jest/jest.builder.spec.ts [hunk 1]
```diff
@@ -4,11 +4,17 @@ jest.mock('jest');
 const { runCLI } = require('jest');
 import * as path from 'path';
 
+class TestJestBuilder extends JestBuilder {
```
Comment (FrozenPandaz) [inline packages/builders/src/jest/jest.builder.spec.ts]: Is this so you can spy? Can you cast to `any` instead please?

File: packages/builders/src/jest/jest.builder.ts [hunk 2]
```diff
@@ -16,37 +17,66 @@ try {
...
+  onlyChanged?: boolean;
+  outputFile?: string;
   passWithNoTests?: boolean;
   runInBand?: boolean;
-  setupFile?: string;
   silent?: boolean;
-  updateSnapshot?: boolean;
   testNamePattern?: string;
+  updateSnapshot?: boolean;
+  useStderr?: boolean;
```
Comment (FrozenPandaz) [inline packages/builders/src/jest/jest.builder.ts]: Can you please revert changes not relevant to `--test-file`?

File: packages/builders/src/jest/jest.builder.ts [hunk 3]
```diff
@@ -16,37 +17,66 @@ try {
...
+    if (options.testFile) {
+      const filePath = options.testDirectory
+        ? path.resolve(options.testDirectory, options.testFile)
+        : realpathSync(options.testFile);
+
+      if (filePath) {
+        const project = this.getProjectForFile(filePath);
+
+        if (project && project.root !== builderConfig.root) {
+          return of({
```
Comment (FrozenPandaz) [inline packages/builders/src/jest/jest.builder.ts]: Why is this necessary? and If the specified file is not in the project, should it be a failure?

File: packages/builders/src/jest/jest.builder.ts [hunk 4]
```diff
@@ -16,37 +17,66 @@ try {
 const { runCLI } = require('jest');
 
 export interface JestBuilderOptions {
+  codeCoverage?: boolean;
   jestConfig: string;
+  testDirectory?: string;
```
Comment (FrozenPandaz) [inline packages/builders/src/jest/jest.builder.ts]: Why is this necessary?

**Review by FrozenPandaz — CHANGES_REQUESTED** [review]
Thank you for making the changes @Domratchev. This PR is looking much better.

**Comment by FrozenPandaz** [comment]
@jasedwards That is unfortunate, can you please create a separate issue?

File: packages/builders/src/jest/jest.builder.spec.ts (hunk 1 above)
Comment (Domratchev) [reply packages/builders/src/jest/jest.builder.spec.ts]: Done.

File: packages/builders/src/jest/jest.builder.ts (hunk 2 above)
Comment (Domratchev) [reply packages/builders/src/jest/jest.builder.ts]: I was initially wrong about the minimum scope of changes required for integration.
All the added properties are used by vscode-jest in different scenarios (some - in debug mode, some - in the background).

File: packages/builders/src/jest/jest.builder.ts (hunk 4 above)
Comment (Domratchev) [reply packages/builders/src/jest/jest.builder.ts]: vscode-jest passes in the file name only, without the full path.
To find out which file exactly that is, and to which project it belongs we need to know the directory.

**Comment by jasedwards** [comment]
@FrozenPandaz  It appears after looking at last commit that watchAll will be available.  can you confirm?  if I am mistaken then I will create a separate issue

**Comment by Domratchev** [comment]
@jasedwards I was my private initiative, hope @FrozenPandaz doesn't mind.

**Comment by Domratchev** [comment]
Also it is very tempting to add passWithNoTests=true by default.
Otherwise testing multiple projects stops when a project with no tests is discovered.

File: packages/builders/src/jest/jest.builder.ts [hunk 5]
```diff
@@ -17,27 +17,47 @@ try {
...
     const options = builderConfig.options;
+
+    if (options.testFile) {
+      const filePath = options.testDirectory
+        ? path.resolve(options.testDirectory, options.testFile)
+        : realpathSync(options.testFile);
+
+      if (filePath && !this._isInFolder(builderConfig.root, filePath)) {
+        return of({
+          success: true
```
Comment (FrozenPandaz) [inline packages/builders/src/jest/jest.builder.ts]: This seems like an error to me. If we look at this out of the context of `vscode-jest` then running the following as a command will pass without errors.

```
> ng test app1 --test-file apps/app2/src/app/app.component.spec.ts
Tests Passed?
```

For VSCode-Jest debugging, it's hard to get the project which is being tested, thus you are testing all of them with your config. Let's consider the following solutions:

1. Compute which project needs to be tested (i don't think VSCode debugging has support for this :()
2. Pass the `testFile` to all projects like you are doing now but also pass `passWithNoTests` so when it doesn't match any files, it will still pass 🤷‍♀️. This solves your use case semantically without causing a false positive for others.
3. Prompt the user for the Project via https://code.visualstudio.com/docs/editor/variables-reference#_input-variables
4. Add configuration to https://github.com/jest-community/vscode-jest to adjust it's logic if the user is using `@nrwl/builders:jest` similar to how it has specific treatment for `create-react-app` projects.

**Review by FrozenPandaz — CHANGES_REQUESTED** [review]
Thank you for your persistence here as we work out the best solution. I am sorry it is taking so long.

Re: `watchAll` I am completely open to adding new options which will improve the tool. However, please do _not_ try to "sneak" changes in. I would have preferred it have it's own issue and it's own PR. It also shows up nicer in the Changelog so other users have a better idea of what changed.

File: packages/builders/src/jest/jest.builder.ts (hunk 5 above)
Comment (Domratchev) [reply packages/builders/src/jest/jest.builder.ts]: @FrozenPandaz Jason, my intent for --test-file was rather specific to vscode-jest integration (or a similar scenario). I did not want to start testing other projects than the one holding the test-file. This way the terminal output is clear. I have also added --pass-with-no-tests to launch.json (see my edited commend above).
Using ng test with --test-file and non-matching project name is not a supported scenario and produces no results (as expected). The hole --test-directory affair is also very much VSCode specific, being the easiest way to provide full path to the file when needed and ignore it when --test-file is empty.

**Comment by Domratchev** [comment]
@FrozenPandaz I do not really like the removal of the project skipping logic. Even with --test-file specified and cwd set it still attempts to find the file in every single project in the workspace. And if the file name is not unique it would find it. I think that logic needs to be there. See my Terminal window output:
cetmac08:formatters adomratchev$ cd /Users/adomratchev/Projects/FARApps/libs/common/src/lib/speech/formatters ; env CI=vscode-jest-tests /usr/local/bin/node --inspect-brk=13240 ../../../../../../node_modules/@angular/cli/bin/ng test --pass-with-no-tests --run-in-band --no-code-coverage --test-file number-formatter.spec.ts --testNamePattern "should recognize plain numbers and ignore spaces" 
Debugger listening on ws://127.0.0.1:13240/71c75eb9-f482-45f6-b376-7d4e57bcc0fb
For help, see: https://nodejs.org/en/docs/inspector
Debugger attached.
No tests found, exiting with code 1
Run with `--passWithNoTests` to exit with code 0
In /Users/adomratchev/Projects/FARApps/apps/count/count
  138 files checked.
  testMatch: /Users/adomratchev/Projects/FARApps/apps/count/count/apps/count/count/**/*.spec.ts, /Users/adomratchev/Projects/FARApps/apps/count/count/libs/services/**.spec.ts, /Users/adomratchev/Projects/FARApps/apps/count/count/libs/alert/**.spec.ts, /Users/adomratchev/Projects/FARApps/apps/count/count/libs/utils/**.spec.ts, /Users/adomratchev/Projects/FARApps/apps/count/count/libs/cet-shared/**.spec.ts - 0 matches
  testPathIgnorePatterns: /node_modules/ - 138 matches
  testRegex:  - 138 matches
Pattern: number-formatter.spec.ts - 0 matches
No tests found, exiting with code 1
Run with `--passWithNoTests` to exit with code 0
In /Users/adomratchev/Projects/FARApps/apps/inspect
  198 files checked.
  testMatch: /Users/adomratchev/Projects/FARApps/apps/inspect/apps/inspect/**/*.spec.ts - 0 matches
  testPathIgnorePatterns: /node_modules/ - 198 matches
  testRegex:  - 198 matches
Pattern: number-formatter.spec.ts - 0 matches
No tests found, exiting with code 1
Run with `--passWithNoTests` to exit with code 0
In /Users/adomratchev/Projects/FARApps/apps/speech-test
  25 files checked.
  testMatch: /Users/adomratchev/Projects/FARApps/apps/speech-test/apps/speech-test/**/*.spec.ts - 0 matches
  testPathIgnorePatterns: /node_modules/ - 25 matches
  testRegex:  - 25 matches
Pattern: number-formatter.spec.ts - 0 matches
No tests found, exiting with code 1
Run with `--passWithNoTests` to exit with code 0
In /Users/adomratchev/Projects/FARApps/libs/auth
  28 files checked.
  testMatch: **/+(*.)+(spec|test).+(ts|js)?(x) - 6 matches
  testPathIgnorePatterns: /node_modules/ - 28 matches
  testRegex:  - 28 matches
Pattern: number-formatter.spec.ts - 0 matches
No tests found, exiting with code 1
Run with `--passWithNoTests` to exit with code 0
In /Users/adomratchev/Projects/FARApps/libs/translate
  6 files checked.
  testMatch: **/+(*.)+(spec|test).+(ts|js)?(x) - 1 match
  testPathIgnorePatterns: /node_modules/ - 6 matches
  testRegex:  - 6 matches
Pattern: number-formatter.spec.ts - 0 matches
No tests found, exiting with code 1
Run with `--passWithNoTests` to exit with code 0
In /Users/adomratchev/Projects/FARApps/libs/cet-core
  8 files checked.
  testMatch: **/+(*.)+(spec|test).+(ts|js)?(x) - 2 matches
  testPathIgnorePatterns: /node_modules/ - 8 matches
  testRegex:  - 8 matches
Pattern: number-formatter.spec.ts - 0 matches
No tests found, exiting with code 1
Run with `--passWithNoTests` to exit with code 0
In /Users/adomratchev/Projects/FARApps/libs/alert
  10 files checked.
  testMatch: **/+(*.)+(spec|test).+(ts|js)?(x) - 3 matches
  testPathIgnorePatterns: /node_modules/ - 10 matches
  testRegex:  - 10 matches
Pattern: number-formatter.spec.ts - 0 matches
No tests found, exiting with code 1
Run with `--passWithNoTests` to exit with code 0
In /Users/adomratchev/Projects/FARApps/apps/fix/fix
  406 files checked.
  testMatch: **/+(*.)+(spec|test).+(ts|js)?(x) - 124 matches
  testPathIgnorePatterns: /node_modules/ - 406 matches
  testRegex:  - 406 matches
Pattern: number-formatter.spec.ts - 0 matches
No tests found, exiting with code 1
Run with `--passWithNoTests` to exit with code 0
In /Users/adomratchev/Projects/FARApps/libs/cet-shared
  33 files checked.
  testMatch: **/+(*.)+(spec|test).+(ts|js)?(x) - 12 matches
  testPathIgnorePatterns: /node_modules/ - 33 matches
  testRegex:  - 33 matches
Pattern: number-formatter.spec.ts - 0 matches
PASS ./number-formatter.spec.ts (187.671s)
  cetNumberFormatter
    ✓ should recognize plain numbers and ignore spaces (185615ms)
    ○ skipped 29 tests

Test Suites: 1 passed, 1 total
Tests:       29 skipped, 1 passed, 30 total
Snapshots:   0 total
Time:        188.786s
Ran all test suites matching /number-formatter.spec.ts/i with tests matching "should recognize plain numbers and ignore spaces".
No tests found, exiting with code 1
Run with `--passWithNoTests` to exit with code 0
In /Users/adomratchev/Projects/FARApps/libs/core
  23 files checked.
  testMatch: **/+(*.)+(spec|test).+(ts|js)?(x) - 6 matches
  testPathIgnorePatterns: /node_modules/ - 23 matches
  testRegex:  - 23 matches
Pattern: number-formatter.spec.ts - 0 matches
Waiting for the debugger to disconnect...
Killed: 9
cetmac08:formatters adomratchev$

**Comment by Domratchev** [comment]
@FrozenPandaz By the way, there is a bug in Jest. It's message about "No tests found, exiting with code 1
Run with --passWithNoTests to exit with code 0" is misleading. That option IS passed in, and Jest knows it, 'cause it proceeds on with testing of subsequent projects, hence the response code IS 0, it just prints the same message regardless.

**Comment by FrozenPandaz** [comment]
Can you fix the commits? Maybe squash them all into a single commit :man_shrugging: 

> By the way, there is a bug in Jest. Its message about "No tests found, exiting with code 1
Run with --passWithNoTests to exit with code 0" is misleading. 

Good Catch :+1: 

> Even with --test-file specified and cwd set it still attempts to find the file in every single project in the workspace. 

This is the nature of using `ng test`. Ideally, it would know to only run the project whose root is the closest parent directory of `process.cwd()` as it does when calling `ng generate`.  :thinking: I reached out to the Angular CLI. It's not the cleanest right now, but we can keep looking for other solutions.

**Comment by Domratchev** [comment]
@FrozenPandaz 
> Can you fix the commits? Maybe squash them all into a single commit 🤷‍♂️
Done. Have a look, please.

**Comment by FrozenPandaz** [comment]
@Domratchev Can you fix the failing tests please?

File: packages/builders/src/jest/jest.builder.ts [hunk 6]
```diff
@@ -56,36 +66,40 @@ export default class JestBuilder implements Builder<JestBuilderOptions> {
...
+      outputFile: options.outputFile,
       passWithNoTests: options.passWithNoTests,
-      silent: options.silent,
       runInBand: options.runInBand,
+      silent: options.silent,
+      testNamePattern: options.testNamePattern,
+      updateSnapshot: options.updateSnapshot,
+      useStderr: options.useStderr,
+      watch: options.watch,
+      wathcAll: options.watchAll,
```
Comment (FrozenPandaz) [inline packages/builders/src/jest/jest.builder.ts]: Typo here

**Review by FrozenPandaz — CHANGES_REQUESTED** [review]

File: packages/builders/src/jest/jest.builder.ts (hunk 6 above)
Comment (Domratchev) [reply packages/builders/src/jest/jest.builder.ts]: Oops. Fixed.

**Comment by lonix1** [comment]
+1

**Review by FrozenPandaz — APPROVED** [review]

**Comment by Karql** [comment]
Hi @FrozenPandaz @Domratchev 

This dose not work on windows:
`    options.jestConfig = path.resolve(
      this.context.workspace.root,
      options.jestConfig
    );`

image

After resolve path is: `C:\C\MDM\nwrltest\ws\apps\todos\jest.config.js` and could not be find
